### PR TITLE
DM-41108: Add analysis_tools afterburner

### PR DESCRIPTION
**Pull request by isullivan**
Adds analysis_tools metrics to the ApPipe and ApPipeWithFakes pipelines.

### CODE REVIEW

File: pipelines/_ingredients/ApPipe.yaml [hunk 1]
```diff
@@ -16,7 +16,8 @@ description: End to end Alert Production pipeline.
 # pipeline, e.g., $AP_PIPE_DIR/pipelines/$CAMERA/ApPipeCalibrateImage.yaml.
 
 imports:
-  - location: $AP_PIPE_DIR/pipelines/_ingredients/ProcessCcd.yaml
+  - $AP_PIPE_DIR/pipelines/_ingredients/ProcessCcd.yaml
+  - $ANALYSIS_TOOLS_DIR/pipelines/apDetectorVisitQualityCore.yaml
```
Comment (kfindeisen) [inline pipelines/_ingredients/ApPipe.yaml:20]: Please add `analysis_tools` to `ups/ap_pipe.table`. Otherwise, setting up `ap_pipe` won't let you use its pipelines.

File: pipelines/_ingredients/ApPipe.yaml [hunk 2]
```diff
@@ -74,6 +75,10 @@ subsets:
       - rbClassify
       - transformDiaSrcCat
       - diaPipe
+  qaMetrics:
+    subset:
+      - analyzeAssocDiaSrcCore
+      - diffimTaskCore
     description: >
       An alias of ApPipe to use in higher-level pipelines.
```
Comment (kfindeisen) [inline pipelines/_ingredients/ApPipe.yaml:83]: The subset definition got a bit garbled; the description belongs with `apPipe`.
```suggestion
    description: >
      An alias of ApPipe to use in higher-level pipelines.
  qaMetrics:
    subset:
      - analyzeAssocDiaSrcCore
      - diffimTaskCore
```

File: pipelines/_ingredients/ApPipe.yaml [hunk 3]
```diff
@@ -74,6 +75,10 @@ subsets:
       - rbClassify
       - transformDiaSrcCat
       - diaPipe
+  qaMetrics:
+    subset:
+      - analyzeAssocDiaSrcCore
+      - diffimTaskCore
```
Comment (kfindeisen) [inline pipelines/_ingredients/ApPipe.yaml:81]: Why not define `qaMetrics` in `apDetectorVisitQualityCore.yaml`? Does `analysis_tools` have rules/conventions against it?

Is this subset supposed to exclude `analyzeTrailedDiaSrcCore`?

Comment (kfindeisen) [reply pipelines/_ingredients/ApPipe.yaml:81]: Please include all tasks in `apPipe` so that the alias behaves correctly in higher-level pipelines like `ApVerify`. Unfortunately, Middleware doesn't want subsets to be included as subset members, so you'll have to do it manually.

File: pipelines/_ingredients/ApPipeWithFakes.yaml [hunk 4]
```diff
@@ -110,6 +111,10 @@ subsets:
       - transformDiaSrcCat
       - diaPipe
       - fakesMatch
+  qaMetrics:
+    subset:
+      - analyzeAssocDiaSrcCore
+      - diffimTaskCore
     description: >
       The AP pipeline with fakes. Requires apPipe and prepareFakes subsets.
```
Comment (kfindeisen) [inline pipelines/_ingredients/ApPipeWithFakes.yaml:119]:
```suggestion
    description: >
      The AP pipeline with fakes. Requires apPipe and prepareFakes subsets.
  qaMetrics:
    subset:
      - analyzeAssocDiaSrcCore
      - diffimTaskCore
```

File: pipelines/_ingredients/ApPipe.yaml (hunk 2 above)
Comment (kfindeisen) [reply pipelines/_ingredients/ApPipe.yaml:83]: If you could add a description for `qaMetrics`, of course, that would be great.

**Review by kfindeisen — CHANGES_REQUESTED** [review]
The subset definitions seem to have gotten mixed up and incomplete; I suspect this may be the root cause of the problems you encountered on lsst/ap_verify#217.

File: pipelines/_ingredients/ApPipeWithFakes.yaml (hunk 4 above)
Comment (kfindeisen) [reply pipelines/_ingredients/ApPipeWithFakes.yaml:119]: The fakes definition of `apPipe` seems to be broken, and to have been previously worked around in the specialized `ApVerifyWithFakes` by using `processCcd` in place of `apPipe`. Not sure if you'd want to fix that, or have me do it, but it would probably be easiest to address it _before_ adding the QA metrics.

Comment (isullivan) [reply pipelines/_ingredients/ApPipeWithFakes.yaml]: I removed the `qaMetrics` subset and put the metrics in `apPipe`. That seems to fix the import problems you pointed out elsewhere. I have verified that a local `ap_verify` run of `ApPipeWithFakes.yaml` on the `ap_verify_ci_hits2015` works.

Comment (kfindeisen) [reply pipelines/_ingredients/ApPipeWithFakes.yaml]: Ok. Just to be clear, I meant for the new tasks to be added to `apPipe` _in addition to_ `qaMetrics`, not instead of. A `qaMetrics` subset might be useful (we don't have one in `ap_verify` because the way metrics are split among different files makes defining it impractical).

File: pipelines/_ingredients/ApPipe.yaml [hunk 5]
```diff
@@ -80,6 +80,14 @@ subsets:
...
     description: >
       An alias of ApPipe to use in higher-level pipelines.
+  promptQaMetrics:
+    subset:
+      - analyzeAssocDiaSrcCore
+      - analyzeTrailedDiaSrcCore
+      - diffimTaskCore
+    description: >
+      QA metrics to run in Prompt Processing.
+
```
Comment (kfindeisen) [inline pipelines/_ingredients/ApPipe.yaml:90]: I don't think you need to redefine this; since you're not excluding any of these tasks on import, the definition from lsst/analysis_tools#218 should carry over.

File: pipelines/_ingredients/ApPipeWithFakes.yaml (hunk 4 above)
Comment (kfindeisen) [reply pipelines/_ingredients/ApPipeWithFakes.yaml]: Filed [DM-43349](https://rubinobs.atlassian.net/browse/DM-43349) to deal with the preexisting bad `apPipe` definition.

**Review by kfindeisen — APPROVED** [review]
Looks clean, thanks!

File: pipelines/_ingredients/ApPipe.yaml (hunk 5 above)
Comment (isullivan) [reply pipelines/_ingredients/ApPipe.yaml:90]: This addition was in response to your comment "A qaMetrics subset might be useful...". Did I misinterpret your request? You are correct that everything works without this extra subset definition.

Comment (kfindeisen) [reply pipelines/_ingredients/ApPipe.yaml:90]: Sorry, when I wrote that I didn't realize you were planning to move (had already moved?) the definition to `analysis_tools`, as previously suggested.

TLDR: it might be useful to have a definition, but there's definitely no need to define it in more than one place.
